File: server/app/routers/redeem.py
from typing import AsyncIterator

from fastapi import APIRouter, Depends, Request
from pydantic import BaseModel

from app.db.base import async_session
from app.services.redeem_repo import RedeemCodeRepo

router = APIRouter()


class RedeemVerifyIn(BaseModel):
    code: str


async def get_redeem_repo() -> AsyncIterator[RedeemCodeRepo]:
    async with async_session() as s:
        yield RedeemCodeRepo(s)


@router.post("/v1/redeem/verify")
async def redeem_verify(req: RedeemVerifyIn, request: Request, repo=Depends(get_redeem_repo)):
    """客户端填买断码激活：验码 + 绑当前设备（X-Device-Id）。
    成功 → {ok:true, product};失败 → {ok:false, reason}（missing_device / invalid_code / device_limit）。
    激活后客户端记买断态、解锁 BYOK——后续翻译走客户端直连，不再经本服务。"""
    device_id = request.headers.get("x-device-id", "")
    if not device_id:
        return {"ok": False, "reason": "missing_device"}
    result = await repo.verify(req.code.strip(), device_id)
    return {"ok": result.ok, "product": result.product, "reason": result.reason}

File: front/lib/redeem.ts
// 买断码激活：客户端填码 → 后端 POST /v1/redeem/verify 验码 + 绑当前设备。
// 成功即在本地记买断态、解锁 BYOK（后续翻译走客户端直连，不再经本服务）。

import { BACKEND_URL } from './config';
import { getDeviceId } from './device';
import { getBuyout, setBuyout } from './storage';

export interface RedeemOutcome {
  ok: boolean;
  product?: string;
  /** 失败原因：missing_device | invalid_code | device_limit | network。 */
  reason?: string;
}

const REASON_TEXT: Record<string, string> = {
  missing_device: '设备标识缺失，请重试',
  invalid_code: '激活码无效或已失效',
  device_limit: '此激活码绑定的设备已达上限',
  network: '无法连接服务器，请检查网络',
};

export function reasonText(reason?: string): string {
  return (reason && REASON_TEXT[reason]) || '激活失败，请重试';
}

/** 提交买断码激活；成功时落地买断态。 */
export async function redeemVerify(code: string): Promise<RedeemOutcome> {
  const trimmed = code.trim();
  if (!trimmed) return { ok: false, reason: 'invalid_code' };
  let deviceId: string;
  try {
    deviceId = await getDeviceId();
  } catch {
    return { ok: false, reason: 'missing_device' };
  }
  let r: Response;
  try {
    r = await fetch(`${BACKEND_URL}/v1/redeem/verify`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', 'X-Device-Id': deviceId },
      body: JSON.stringify({ code: trimmed }),
    });
  } catch {
    return { ok: false, reason: 'network' };
  }
  if (!r.ok) return { ok: false, reason: 'network' };
  const data = (await r.json().catch(() => ({}))) as RedeemOutcome;
  if (data.ok) {
    const prev = await getBuyout();
    await setBuyout({ ...prev, active: true, code: trimmed, activatedAt: Date.now() });
  }
  return data;
}
